cli/client.go

```diff
@@ -425,6 +426,123 @@ func RobotsPartShellAction(c *cli.Context) error {
+}
+
+// MachinesPartCopyFilesAction is the corresponding Action for 'machines part cp'.
+func MachinesPartCopyFilesAction(c *cli.Context) error {
+	args := c.Args().Slice()
+	if len(args) == 0 {
+		return errNoFiles
+	}
+
+	client, err := newViamClient(c)
+	if err != nil {
+		return err
+	}
+
+	// Create logger based on presence of debugFlag.
+	logger := logging.FromZapCompatible(zap.NewNop().Sugar())
+	if c.Bool(debugFlag) {
+		logger = logging.NewDebugLogger("cli")
+	}
+
+	// the general format is
+	// from:
+	//		cp machine:path1 machine:path2 ... local_destination
+	// to:
+	//		cp path1 path2 ... remote_destination
+	// we just need to look for machine: to determine what the user's intent is
+	determineDirection := func(args []string) (isFrom bool, destination string, paths []string, err error) {
+		const machinePrefix = "machine:"
+		isFrom = strings.HasPrefix(args[0], machinePrefix)
+
+		if strings.HasPrefix(args[len(args)-1], machinePrefix) {
+			if isFrom {
+				return false, "", nil, errLastArgOfFromMissing
+			}
+		} else if !isFrom {
+			return false, "", nil, errLastArgOfToMissing
+		}
+
+		destination = args[len(args)-1]
+		if !isFrom {
+			destination = strings.TrimPrefix(destination, machinePrefix)
+		}
+
+		// all but the last arg are what we are copying to/from
+		for _, arg := range args[:len(args)-1] {
+			if isFrom && !strings.HasPrefix(arg, machinePrefix) {
+				return false, "", nil, copyFromPathInvalidError{arg}
+			}
+			if isFrom {
+				arg = strings.TrimPrefix(arg, machinePrefix)
+			}
+			paths = append(paths, arg)
+		}
+		return
+	}
+
+	isFrom, destination, paths, err := determineDirection(args)
+	if err != nil {
+		return err
+	}
+
+	doCopy := func() error {
+		if isFrom {
+			return client.copyFilesFromMachine(
+				c.String(organizationFlag),
+				c.String(locationFlag),
+				c.String(machineFlag),
+				c.String(partFlag),
+				c.Bool(debugFlag),
+				c.Bool(cpFlagRecursive),
+				c.Bool(cpFlagPreserve),
+				paths,
+				destination,
+				logger,
+			)
+		}
+
+		return client.copyFilesToMachine(
+			c.String(organizationFlag),
+			c.String(locationFlag),
+			c.String(machineFlag),
+			c.String(partFlag),
+			c.Bool(debugFlag),
+			c.Bool(cpFlagRecursive),
+			c.Bool(cpFlagPreserve),
+			paths,
+			destination,
+			logger,
+		)
+	}
+	if err := doCopy(); err != nil {
+		if statusErr := status.Convert(err); statusErr != nil &&
+			statusErr.Code() == codes.InvalidArgument &&
+			statusErr.Message() == shell.ErrMsgDirectoryCopyRequestNoRecursion {
+			return errors.New(errMsgDirectoryCopyRequestNoRecursionForCLI)
+		}
+		return err
+	}
+	return nil
+}
+
+// errMsgDirectoryCopyRequestNoRecursionForCLI should be returned when a file is included in a path for a copy request
+// where recursion is not enabled.
+var errMsgDirectoryCopyRequestNoRecursionForCLI = "file is a directory but copy recursion not used (you can use -r to enable this)"
```

Comment:
might as well just make the error here instead of making the var a string. Also could just lump with the errors at the top of the file